### ISSUE TITLE
How to define density values based on the pressure and temperature of a particular particle?

### ISSUE THREAD

**Issue by Prasanna2989** [underworldcode/underworld2#446]
Hi All,

I am trying to apply a particular density value to a particular particle.  I want to check the static pressure and the temperature of the marker and if `(temperature == certain value)` and (pressure=>some value), assign a particular density value `(rho0*(1.-alpha*deltaT)*(1+0.4*Mext))` to that particle. Here `Mext` is a swarm. How can I map this new function to a particular particle?  variable.

```python
static_pressure = h*3300.*g
pr = static_pressure.evaluate(particle coordinate)
T = temperatureField.evaluate(particle coordinate)

if (pr>=some value) & (T==certain value):
    #The following segment is to show you the idea
    densityFn(particle coordinate) = (rho0(1.-alpha*deltaT)(1+0.4*Mext))
```
Please let me know if you need further clarification for the question.

Thanks
Pras

**Comment by jmansour** [comment]
Firstly, let me point you towards github syntax highlighting so that you can make your posts more readable:
https://help.github.com/en/github/writing-on-github/creating-and-highlighting-code-blocks
I've updated your post accordingly 🙂
I think you'll want to look at the `underworld.function.branching.conditional()` function:
https://underworld2.readthedocs.io/en/latest/build/underworld.function.branching.html#underworld.function.branching.conditional
https://underworld2.readthedocs.io/en/latest/build/user_guide/04_Functions.html

Something like this:
```python
delta = 0.001  # you'll have to decide an appropriate delta
press_condition = static_pressure >= p_val
temp_condition = (temperatureField < t_val+delta) & (temperatureField > t_val-delta)
dens_fn1 = (rho0(1.-alpha*deltaT)(1+0.4*Mext))  # Mext doesn't really make sense here.. 
dens_fn2 = ?  # what do you want otherwise? 
densityFn = fn.branching.conditional(   ( (press_condition & temp_condition), dens_fn1),
                                                                              (True, dens_fn2)  )  )  
```
Alternatively, you might simply create a density `double` type swarm variable and do a loop over all particles in Python updating as necessary.

**Comment by Prasanna2989** [comment]
Thank you jmansour,

I have two questions regarding your answer and I have quoted your answer sections for further clarity.

1) jmansour  Answer- 
```dens_fn1 = (rho0(1.-alpha*deltaT)(1+0.4*Mext))  # Mext doesn't really make sense here..```

The 'Mext' is a swarm variable and the density of a particular marker has to be increased when the value of Mext is increased with the model time steps. (The Mext is the amount of melt extraction from a particular marker in the Geological perspective (Sizova et al., 2015))


2) jmansour  answer- "Alternatively, you might simply create a density double type swarm variable and do a loop over all particles in Python updating as necessary."

If I follow this method and create a density swarm variable, how should I give the 'buoyancyFn' in the following code segment.

```
stokes = uw.systems.Stokes( velocityField = velocityField, 
                            pressureField = pressureField,
                            conditions    = freeslipBC,
                            fn_viscosity  = viscosityFn, 
                            fn_bodyforce  = buoyancyFn )```

Thank you
Pras

**Comment by jmansour** [comment]
Hi Pras

1.  Ok, got it, that's fine then. Note that in your original post you said that `Mext` was a swarm. 
2.  You'd do something like this

```python
dens_var = swarm.add_variable("double",1)
for index,coord in enumerate(swarm.data):
    temp = temperatureField.evaluate(coord)
    press = static_pressure.evaluate(coord)
    if (temp>something) and (press<something_else):
        dens_var.data[index] = some_dens(coord)
    else:
        dens_var.data[index] = some_other_dens(coord)
buoyancyFn = (1- 0.01*dens_var)*gravFn
stokes = uw.systems.Stokes( velocityField = velocityField, 
                            pressureField = pressureField,
                            conditions    = freeslipBC,
                            fn_viscosity  = viscosityFn, 
                            fn_bodyforce  = buoyancyFn )
```

**Comment by Prasanna2989** [comment]
Thanks John. It helped a lot. 

However, I have an issue when I was trying to apply the same approach in your first answer with a swarm variable value. 

For an example, in your first answer, the conditions are given as follows.
```
temp_condition = (temperatureField < t_val+delta)
```
I used the same approach to apply the conditions on a fn.branching.conditional based on a swarm variable (MatWeakIndex) as per following code segment.

```
MatWeakIndex   = swarm.add_variable( dataType="int", count=1 )

Arr = 0
weak = 1
MatWeakIndex.data[:] = weak

for index, position in enumerate(swarm.particleCoordinates.data):
    if position[1] > 0.5:
        MatWeakIndex.data[index] = weak
    else:
        MatWeakIndex.data[index] = Arr
               
con_weak1 = (MatWeakIndex == weak)
fn_conditional = fn.branching.conditional( [ (con_weak1, tauY_ductile_Weaken_melt), (True, tauY_ductile) ])
#tauY_ductile_Weaken_melt,  tauY_ductile are two functions
```

This is not working properly and directly execute the default function (True, tauY_ductile) instead of evaluating (con_weak1, tauY_ductile_Weaken_melt). May be con_weak1 goes to False in all the cases.

I was wandering whether the conditions for swarm variable act in the same way as a mesh variable or is it different. If it is different how should I give the following condition in this case?

```
con_weak1 = (MatWeakIndex == weak)
```

Thanks
Pras

**Comment by jmansour** [comment]
If you are doing things by index, you'll need to use `fn.branching.map()`.

Due to difficulty with comparing floats, I believe we don't actually overload `==`, and therefore your statement `MatWeakIndex==weak` probably evaluates directly to `False`, hence your observed behaviour.  

Check in Python what `MatWeakIndex==weak` is.

**Comment by Prasanna2989** [comment]
Yes ```MatWeakIndex==weak``` evaluated as False for all the cases.

The problem of using fn.branching.map() is, I have 2 different indexing in this scenario.
For an example 

the first indexing is 
```
MatWeakIndex   = swarm.add_variable( dataType="int", count=1 )

Arr = 0
weak = 1
MatWeakIndex.data[:] = weak

for index, position in enumerate(swarm.particleCoordinates.data):
    if position[1] > 0.5 & position[1] > 0.2 & position[1] < 0.5:
        MatWeakIndex.data[index] = weak
    else:
        MatWeakIndex.data[index] = Arr
        
con_weak1 = (MatWeakIndex == weak)    # issue exists with this part   
fn_conditional = fn.branching.conditional( [ (con_weak1, tauY_ductile_Weaken_melt), (True, tauY_ductile) ])        

viscosity_Y_M = fn_conditional/(2.*(strainRate_2ndInvariant+1.0e-18))
```

The following mapping cannot be removed because I need to apply densityFn based on following material indices.

second indexing
```
viscosityMap = { MantleIndex : viscosity_Y_M,  
                 wetbasaltIndex : viscosity_Y_M,  
                 drybasaltIndex : viscosity_Y_M,  
                 meltBearPeri : viscosity_Y_M,
                 meltBearDepPeri : viscosity_Y_M,
                 depPeri : viscosity_Y_Mr,
                 felsic_melt_DB : viscosity_Y_M,
                 felsic_melt_WB : viscosity_Y_M,
                 basaltic_melt_wb: viscosity_Y_M} 
viscosityFn = fn.branching.map( fn_key = materialVariable, mapping = viscosityMap )
```

I can use a solution like below to overcome this issue of indexing.

```
meltCoords = np.array([ [0.2,0.2], [0.4,0.4],[0.6,0.6]])

dx = 0.5

coord = fn.input()

condition1 = ((coord[1]>swarmCoords[:,1][0]) & (coord[0]> (swarmCoords[:,0][0]-dx)) & (coord[0]<(swarmCoords[:,0][0]+dx)))
condition2 = ((coord[1]>swarmCoords[:,1][1]) & (coord[0]> (swarmCoords[:,0][1]-dx)) & (coord[0]<(swarmCoords[:,0][1]+dx)))
condition3 = ((coord[1]>swarmCoords[:,1][2]) & (coord[0]> (swarmCoords[:,0][2]-dx)) & (coord[0]<(swarmCoords[:,0][2]+dx)))

condition_Weak  = condition1 | condition2 | condition3
fn_conditional = fn.branching.conditional( [ (condition_Weak, tauY_ductile_Weaken_melt), (True, tauY_ductile ) ])
```

but the problem is, the meltCoords array changed in each time steps and I am not sure how to update the conditions in each time step.

Basically what I want to do is, I want to apply a weaker rheology(viscosity) in the blue region in this attachment.
[issue.pptx](https://github.com/underworldcode/underworld2/files/3996578/issue.pptx)
